api/src/services/schema.ts

```diff
@@ -0,0 +1,290 @@
+			})
+		),
+		relations: Joi.array().items(
+			Joi.object({
+				hash: Joi.string().when('diff.0.kind', { is: 'N', then: Joi.optional(), otherwise: Joi.any() }),
+				collection: Joi.string(),
+				field: Joi.string(),
+				related_collection: Joi.any(),
+				diff: Joi.array().items(deepDiffSchema),
+			})
+		),
+	}).required(),
+});
+
+export class SchemaService {
+	knex: Knex;
+	accountability: Accountability | null;
+
+	constructor(options: Omit<AbstractServiceOptions, 'schema'>) {
+		this.knex = options.knex || getDatabase();
+		this.accountability = options.accountability || null;
+	}
+
+	async snapshot(): Promise<Snapshot> {
+		if (this.accountability?.admin !== true) throw new ForbiddenException();
+
+		const currentSnapshot = await getSnapshot({ database: this.knex });
+
+		return currentSnapshot;
+	}
+
+	async apply(payload: { hash: string; diff: SnapshotDiff }): Promise<void> {
+		if (this.accountability?.admin !== true) throw new ForbiddenException();
+
+		const { error } = applyJoiSchema.validate(payload);
+		if (error) throw new InvalidPayloadException(error.message);
+
+		const currentSnapshot = await this.snapshot();
+
+		const snapshotWithHash = this.getHashedSnapshot(currentSnapshot);
+
+		if (payload.hash !== snapshotWithHash.hash) {
+			for (const diffCollection of payload.diff.collections) {
+				const collection = diffCollection.collection;
+
+				if (diffCollection.diff[0]?.kind === 'N') {
+					const existingCollection = snapshotWithHash.collections.find(
+						(c) => c.collection === diffCollection.collection
+					);
+					if (existingCollection) {
+						throw new InvalidPayloadException(
+							`Provided diff is trying to create collection "${collection}" but it already exists. Please regenerate a new diff and try again.`
+						);
+					}
+				} else if (diffCollection.diff[0]?.kind === 'D') {
+					const existingCollection = snapshotWithHash.collections.find(
+						(c) => c.collection === diffCollection.collection
+					);
+					if (!existingCollection) {
+						throw new InvalidPayloadException(
+							`Provided diff is trying to delete collection "${collection}" but it does not exist. Please regenerate a new diff and try again.`
+						);
+					}
+				} else {
+					const matchingCollection = snapshotWithHash.collections.find((c) => c.hash === diffCollection.hash);
+					if (!matchingCollection) {
+						throw new InvalidPayloadException(
+							`Provided diff to update collection "${collection}" does not match the current instance's collection, indicating it has changed after this diff was generated. Please regenerate a new diff and try again.`
+						);
+					}
+				}
+			}
+			for (const diffField of payload.diff.fields) {
+				const field = `${diffField.collection}.${diffField.field}`;
+
+				if (diffField.diff[0]?.kind === 'N') {
+					const existingField = snapshotWithHash.fields.find(
+						(f) => f.collection === diffField.collection && f.field === diffField.field
+					);
+					if (existingField) {
```

Comment:
Tiny tiny nitpick, but I personally prefer putting an empty line between blocks like these. Makes the whole thing just a tad more readable